### ISSUE TITLE
Use tags instead of branches for releases

### ISSUE THREAD

**Issue by rcdailey** [powervr-graphics/Native_SDK#27]
When using Git, tags should be used to mark specific commits as versioned releases of your software. However, right now branches are being used to represent those releases. In addition, a `master` branch should be provided that is continuously updated and always represents the latest release. Specifically, the following changes should be made:

1. Convert every numbered release branch to a corresponding tag (e.g. `refs/heads/5.0` will become `refs/tags/v5.0`).
1. Create `master` branch from the latest current release (at the time of this report, that is 5.2)
1. Set `master` as the default branch in repo settings
1. Delete all numbered branches in `refs/heads`.

I have already done this work on my fork: https://github.com/rcdailey/Native_SDK

Please review it and let me know if you find these changes acceptable. If you accept these changes, your releases will be more streamlined and more idiomatic to Git.

**Comment by rcdailey** [comment]
Since it seems this repository is only used to distribute releases and not for active development, then for each release your process will be:

1. Check out `master`
1. Delete all files in your working copy except `.git`
1. Copy/extract the contents of your complete release into your working copy
1. Do `git add .` to add all new files, modified files, and deleted files
1. Do `git commit` and leave an appropriate message
1. Do `git tag -a v1.0` to create an annotated tag (named `v1.0`, but replace this with the actual release number). You will be prompted for a message to describe the tag (You can put release information here, such as maybe release notes, or other important info). The tag is created at `HEAD` (the commit you just made).
1. Push the tag and new commit with `git push origin v1.0 master`. This pushes the new tag (again, replace `v1.0` with your actual tag name) and `master`, which contains the commit for the new release you just made.

Doing things this way will also reduce the size of your repository, assuming that between releases there's a common base and you're modifying existing files.

**Comment by PVRDevTech** [comment]
Hi rcdailey,

A branch-per-release strategy (at least similar to what we are following at the moment) is -at least for the moment - necessary because we sometimes have to do changes/fixes to previous branches, and some of the releases (especially the major ones) are different enough that some users may not planning to upgrade at all (as an example, the 4.x branch uses a completely different API abstraction to 5.x, so would be more than a tag. As such, bugfixes that do not apply to the subsequent mainline may be possible to be required for it.

We are planning to move to something similar, but not exactly the same as what you are proposing, namely to at least keep a live master branch on Github. There are various things to consider and solve before this is possible though (infrastructure necessary to conform to some policies), so I can't really give a timeline for it, but this is definitely under consideration,

Kind regards,
Gerry

**Comment by PVRDevTech** [comment]
We will take this under very careful consideration.
What you are saying is very close to where we want to be. But to be honest at that point the two solutions we have under consideration (What you say:"master is only branch, releases are tags" and what we plan: "master is mainline, releases branch off") have little enough difference to be academic (the main reason we are thinking of two is that sometimes we get even feature requests for old releases, and there are differences enough that independend development of, say, 4.0 is not inconceivable). In your opinion though - wouldn't the large part of the benefit still be the use of the (conspicuously missing now) master branch?

**Comment by graptis** [comment]
Hi Robert,
There has been a lot of effort on this one, mainly due to making Git do what we want it to you. I just wanted to get your opinion on mechanics of things and if potential benefits outweigh the potential pain.

@omarzohdi (IMG employee as well) is currently almost done with this and he is trying to 1) tie all the orphan branches to the mainline, and 2) clean up the tree as much as humanly possible.

The idea here is that quite some amount of "history rewriting" will need to happen for us to get to a sensible result, but the idea is for us to rebase all the orphan branches on the mainline (I think the longest running is 4.3), tag the releases as necessary and then either delete or merge back the branches.

This should allow us to clean things completely, and the releases will be tagged.

In short, based on your feedback (and other issues) we'll go the "simple" way: Tag per release, hotfix branches as required, but I wanted to ask if (at least for you but a sample of 1 is much better than a sample of 0) the aforementioned changes would cause you undue pain due to the amount of history rewriting that will need to happen.

**Comment by graptis** [comment]
Continued:
The other option we have is do what I mentioned before, but on a copy of whatever branch we keep as mainline. This will make the repo look much less clean, but has a much greater opportunity of not causing pain to people.
In that case, we would create a new single master containing onto which we will rebase all the other "old, stale" branches, tag the releases on that, and delete all the other branches. The benefit there of course is that no history rewriting will need to take place.

**Comment by rcdailey** [comment]
Hey guys, good to hear from you again. It's good to hear you are cleaning this up. It is great for me, but your team will benefit the most from it because it will make release management much smoother and easier for you.

I read back through our conversation again to refresh my memory since it's been a while since we last spoke. The first topic I'd like to address is your intention to go back and retroactively fix your current/past release branches. While you can certainly do this, I feel obligated to caution you on one obvious risk of this. Depending on the way you do this, and the state of the code, you could inadvertently modify previous releases. For example, if you do this and you end up resolving conflicts or unintentionally introducing other changes, you won't be able to 100% replicate a previous release. Modifying the code to some extent is unavoidable because you'd be doing merges to connect release mainlines together. If you don't like the risk of say, a v4.2 release potentially being different between two timelines (before your `filter-branch` and after it), I would avoid doing that. From a business standpoint, you probably don't have much to gain from fixing previous release branches anyway.

I am not discouraging doing this so much as just making the pitfalls known. If you can do this and the binary end result of your releases doesn't change significantly or at all, you should be fine. And any auditing you do on your code history will benefit from it.

My original suggestion was to leave previous release branches orphaned. This was to avoid the risk of the issues I mentioned before, as well as to save you unnecessary work. As long as you start doing things the "new way" with your latest release branch (previously I had given an example of creating `master` off of `5.2`), it should get you on track.

To your question about how this work would affect me: It wouldn't. But that's just me. I don't know how many other folks have forks and if they have significant work or not. But it shouldn't affect them either. Ultimately, you're not changing the actual code. Just the structure of the repository. If you completely rewrite it, someone should be able to do a simple `git rebase --onto` to rebase their feature branches back onto your new mainline branches and there will be no conflicts. So for myself and everyone else, I can say that this is safely a non-issue. Git rebase is designed to handle cases like this. Obviously you don't want to do it often, but given the reasons for it I think it's perfectly acceptable.

A third potential option is to rename this repository to `Native_SDK_Legacy` and create a new one called `Native_SDK`, where you push your new rewritten repository. `Native_SDK_Legacy` would then become read-only. If users are affected by your change, you can add a README that instructs them to change their `origin` remote to the legacy repo, and they are back to where they were. Then at their own time, they can use `rebase --onto` to migrate their branches to the new repository (via another remote in their clone).

**Comment by rcdailey** [comment]
Also: You mentioned some struggles with Git to figure out how to migrate all your branches together. Your rebase idea works out, but you could also just start off replicating your whole repository again from scratch since you only have 17 commits.

If you'd like my help, let me know. I'm very familiar with rewriting repositories with Git.

**Comment by graptis** [comment]
Hi

**Comment by graptis** [comment]
Hi @rcdailey , we just pushed a brand-new stitched-together mainline branch and our releases are now tags. For the moment we are leaving the legacy branches around to avoid disruption.

**Comment by omarzohdi** [comment]
As mentioned by @graptis we have a single default master branch and tags for each release. Legacy branches will be deleted in the future but for now we'll leave them around to keep forks in sync. 

Closing the issue.
